python/dgl/distributed/rpc_server.py

```diff
@@ -63,39 +67,52 @@ def start_server(server_id, ip_config, num_servers, num_clients, server_state, \
     # wait all the senders connect to server.
     # Once all the senders connect to server, server will not
     # accept new sender's connection
-    print("Wait connections non-blockingly...")
+    print(
+        "Server is waiting for connections non-blockingly on [{}:{}]...".format(ip_addr, port))
     rpc.receiver_wait(ip_addr, port, num_clients, blocking=False)
     rpc.set_num_client(num_clients)
-    # Recv all the client's IP and assign ID to clients
-    addr_list = []
-    client_namebook = {}
-    for _ in range(num_clients):
-        # blocked until request is received
-        req, _ = rpc.recv_request()
-        assert isinstance(req, rpc.ClientRegisterRequest)
-        addr_list.append(req.ip_addr)
-    addr_list.sort()
-    for client_id, addr in enumerate(addr_list):
-        client_namebook[client_id] = addr
-    for client_id, addr in client_namebook.items():
-        client_ip, client_port = addr.split(':')
-        # TODO[Rhett]: server should not be blocked endlessly.
-        while not rpc.connect_receiver(client_ip, client_port, client_id):
-            time.sleep(1)
-    if rpc.get_rank() == 0: # server_0 send all the IDs
-        for client_id, _ in client_namebook.items():
-            register_res = rpc.ClientRegisterResponse(client_id)
-            rpc.send_response(client_id, register_res)
-    # main service loop
+    recv_clients = {}
     while True:
-        req, client_id = rpc.recv_request()
+        # go through if any client group is ready for connection
+        for group_id in list(recv_clients.keys()):
```

Comment:
why do you put the code of handling client registration code here? it looks weird.

Reply:
as this is where we try to connect peer clients for the same group. It's business logic and I think it's better to put here than in underlying c++ libs. In the future, I prefer to repalce sender/receiver with client/server. Once connection arrives, it's bi-directional, we do not need to create a sender for sending request.